File: 백준/21608/aeri.py
n = int(input())
studentList = [list(map(int, input().split())) for i in range(n*n)]
myMap = [[0] * n for i in range(n)]

move = [[-1, 0], [1, 0], [0, 1], [0, -1]]

answer = 0


def find_empty(student):
    queue = []
    for i in range(n):
        for j in range(n):
            if(myMap[i][j] == 0):
                countEmpty = 0
                countFavorite = 0
                for dir in range(4):
                    nx, ny = i + move[dir][0], j + move[dir][1]
                    if(nx >= 0 and nx < n and ny >= 0 and ny < n):
                        if(myMap[nx][ny] == 0):
                            countEmpty += 1
                        elif myMap[nx][ny] in student[1:]:
                            countFavorite += 1
                queue.append([countFavorite, countEmpty, i, j])

    queue.sort(reverse=True)
    emptyMax = queue[0][0]
    queue.sort()
    favoriteMax = 0
    for i in range(len(queue)):
        if(queue[i][0] == emptyMax):
            if(favoriteMax < queue[i][1]):
                favoriteMax = queue[i][1]
    for i in range(len(queue)):
        if(queue[i][0] == emptyMax and queue[i][1] == favoriteMax):
            myMap[queue[i][2]][queue[i][3]] = student[0]
            return


for i in range(len(studentList)):
    find_empty(studentList[i])

studentList.sort()
for i in range(n):
    for j in range(n):
        count = 0
        for dir in range(4):
            nx, ny = i+move[dir][0], j + move[dir][1]
            if(nx >= 0 and nx < n and ny >= 0 and ny < n):
                # print("studentList[myMap[i][j]-1]", studentList[myMap[i][j]-1])
                if(myMap[nx][ny] in studentList[myMap[i][j]-1][1:]):
                    count += 1
        # print("count", count)
        if(count == 4):
            answer += 1000
        elif(count == 3):
            answer += 100
        elif(count == 2):
            answer += 10
        elif(count == 1):
            answer += 1
print(answer)

File: 백준/21608/aeri.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int myMap[20][20];
int moveXY[4][2] = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};

void checkPosition(int n, vector<int> student) {
    vector<vector<int>> queue;
    int emptyMax = 0, favoriteMax = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (myMap[i][j] == 0) {
                int countEmpty = 0, countFavorite = 0;
                for (int dir = 0; dir < 4; dir++) {
                    int nx, ny;
                    nx = i + moveXY[dir][0];
                    ny = j + moveXY[dir][1];
                    if (nx >= 0 && nx < n && ny >= n && ny < n) {
                        if (myMap[nx][ny] == 0) {
                            countEmpty++;
                        }

                        else {
                            auto tmp = find(student.begin(), student.end(), myMap[nx][ny]);
                            if (tmp != student.end()) {
                                countFavorite++;
                            }
                        }
                    }
                }
                queue.push_back({countEmpty, countFavorite, i, j});
            }
        }
    }
    sort(queue.begin(), queue.end(), greater<int>());
    emptyMax = queue[0][0];
    sort(queue.begin(), queue.end());

    for (int i = 0; i < queue.size(); i++) {
        if (queue[i][0] == emptyMax) {
            if (favoriteMax < queue[i][1]) {
                favoriteMax = queue[i][1];
            }
        }
    }
    for (int i = 0; i < queue.size(); i++) {
        if (emptyMax == queue[i][0] && favoriteMax == queue[i][1]) {
            myMap[queue[i][2]][queue[i][3]] = student[0];
            return;
        }
    }
}

int main() {
    int n, answer = 0;
    cin >> n;
    vector<vector<int>> studentList(n * n, vector<int>(5, 0));

    for (int i = 0; i < n * n; i++) {
        cin >> studentList[i][0] >> studentList[i][1] >> studentList[i][2] >> studentList[i][3] >> studentList[i][4];
    }

    for (int i = 0; i < n * n; i++) {
        checkPosition(n, studentList[i]);
    }
    sort(studentList.begin(), studentList.end());
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int count = 0;
            for (int dir = 0; dir < 4; dir++) {
                int nx, ny;
                nx = i + moveXY[dir][0];
                ny = j + moveXY[dir][1];
                if (nx >= 0 && nx < n && ny >= n && ny < n) {
                    auto tmp = find(studentList[myMap[i][j] - 1].begin(), studentList[myMap[i][j] - 1].end(), myMap[nx][ny]);
                    if (tmp != studentList[myMap[i][j] - 1].end()) {
                        count++;
                    }
                }
            }
            if (count == 4) {
                answer += 1000;
            } else if (count == 3) {
                answer += 100;
            } else if (count == 2) {
                answer += 10;
            } else if (count == 1) {
                answer += 1;
            }
        }
    }
    cout << answer;
}